File: examples/ios-sdk-examples/ios-sdk-examples-objc/Views/Polygon/PolygonViewController.m
//
//  PolygonViewController.m
//  ios-sdk-examples-objc
//
//  Created by iMacbook on 2/2/21.
//

#import "PolygonViewController.h"
#import <Map4dMap/Map4dMap.h>

@interface PolygonViewController ()

@property (weak, nonatomic) IBOutlet MFMapView *mapView;
@property (weak, nonatomic) IBOutlet UILabel *titleLabel;
@property (weak, nonatomic) IBOutlet UIButton *removeButton;
@end

@interface PolygonViewController(MFMapViewDelegate) <MFMapViewDelegate>
@end

@implementation PolygonViewController {
  MFPolygon *polygon;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  [[self navigationItem] setTitle: @"POLYGON"];
  [_mapView setPOIsEnabled: NO];
  [_mapView setDelegate: self];
  polygon = [[MFPolygon alloc] init];
  [self draw: polygon];
}

- (IBAction)didTapAdd:(id)sender {
  [_removeButton setHidden: NO];
  [_titleLabel setHidden: YES];
  if (polygon != nil) {
    return;
  }
  polygon = [[MFPolygon alloc] init];
  [self draw:polygon];
}

- (IBAction)didTapRemove:(id)sender {
  [_titleLabel setHidden: YES];
  [_removeButton setHidden: YES];
  [polygon setMap: nil];
  polygon = nil;
}

- (void) draw: (MFPolygon*) polygon {
  MFMutablePath *path = [[MFMutablePath alloc] init];
  [path addCoordinate:CLLocationCoordinate2DMake(16.072699584731552, 108.23020984691385)];
  [path addCoordinate:CLLocationCoordinate2DMake(16.07608106276882, 108.22671224150031)];
  [path addCoordinate:CLLocationCoordinate2DMake(16.0730244298952, 108.22336243622124)];
  [path addCoordinate:CLLocationCoordinate2DMake(16.070900670809152, 108.22593736045201)];
  [path addCoordinate:CLLocationCoordinate2DMake(16.070485456927244, 108.22957624088468)];
  [path addCoordinate:CLLocationCoordinate2DMake(16.072699584731552, 108.23020984691385)];

  MFMutablePath *hole = [[MFMutablePath alloc] init];
  [hole addCoordinate:CLLocationCoordinate2DMake(16.073472611109437, 108.22747377269246)];
  [hole addCoordinate:CLLocationCoordinate2DMake(16.071575666602996, 108.22781709595301)];
  [hole addCoordinate:CLLocationCoordinate2DMake(16.072503522154975, 108.22573569886646)];
  [hole addCoordinate:CLLocationCoordinate2DMake(16.073946844408653, 108.22612193750109)];

  NSMutableArray *holes = [[NSMutableArray alloc] init];
  [holes addObject: hole];

  [polygon setPath: path];
  [polygon setHoles: holes];
  [polygon setStrokeWidth:20.0];
  [polygon setStrokeColor:[self randomColor]];
  [polygon setFillColor: [self randomColor]];
  [polygon setMap: _mapView];
}

- (UIColor *)randomColor
{
  double red = random() % 255 / 255.0;
  double green = random() % 255 / 255.0;
  double blue = random() % 255 / 255.0;
  UIColor *color = [UIColor colorWithRed:red green:green blue:blue alpha:1.0];
  NSLog(@"%@", color);
  return color;
}
@end

@implementation PolygonViewController(MFMapViewDelegate)
- (void)mapview:(MFMapView *)mapView didTapPolygon:(MFPolygon *)polygon {
  [_titleLabel setHidden: NO];
  [_titleLabel setText: [NSString stringWithFormat: @"PolygonId: %u", (unsigned int)polygon.Id]];
}

@end

File: examples/ios-sdk-examples/ios-sdk-examples/Views/Home/polygons/PolygonViewController.swift
//
//  PolygonViewController.swift
//  ios-sdk-examples
//
//  Created by tantv on 1/9/19.
//  Copyright © 2019 tantv. All rights reserved.
//

import UIKit
import map4dsdk

class PolygonViewController: UIViewController {
  @IBOutlet weak var lbTitle: UILabel!
  
  @IBOutlet weak var lbHide: UIButton!
  @IBOutlet weak var map4d: MFMapView!
  var pathPolygon = MFMutablePath.init()
  var hole = MFMutablePath.init()
  var isInitPolygon = false
  var polygon: MFPolygon?
  var isAddPolygon = false
  override func viewDidLoad() {
        super.viewDidLoad()
      self.navigationItem.title = "Polygon"
      addPolygon()
    map4d.delegate = self
  }
  override func viewDidLayoutSubviews() {
    
  }
  func addPolygon(){
    if (!isInitPolygon){
      initData()
    }
    if (!isAddPolygon){
      
      // Move Camera To Polygon Position Before Draw
      let bounds = MFCoordinateBounds().includingPath(pathPolygon)
      let cameraUpdate = MFCameraUpdate.fit((bounds as! MFCoordinateBounds))
      map4d.moveCamera(cameraUpdate)
      
      polygon = MFPolygon.init(pathPolygon, holes: [hole])
      polygon?.map = map4d
      isAddPolygon = true
    }
  }
  func initData(){
    if (!isInitPolygon){
      pathPolygon.removeAllCoordinates()
      // Notice important
      // First point = End Point
      pathPolygon.add(CLLocationCoordinate2DMake(16.061583,108.234463))
      pathPolygon.add(CLLocationCoordinate2DMake(16.038148,108.243031))
      pathPolygon.add(CLLocationCoordinate2DMake(16.032951,108.224449))
      pathPolygon.add(CLLocationCoordinate2DMake(16.031512,108.208902))
      pathPolygon.add(CLLocationCoordinate2DMake(16.061583,108.234463))
      // First point = End Point
      // Init hole
      // Notice important
      // First point = End Point
      hole.removeAllCoordinates()
      hole.add(CLLocationCoordinate2DMake(16.043380,108.227177))
      hole.add(CLLocationCoordinate2DMake(16.036575,108.233614))
      hole.add(CLLocationCoordinate2DMake(16.034430,108.220396))
      hole.add(CLLocationCoordinate2DMake(16.040122,108.220096))
      hole.add(CLLocationCoordinate2DMake(16.043380,108.227177))
      // First point = End Point
      
      isInitPolygon = true
    }
  }
  @IBAction func add(_ sender: Any) {
    addPolygon()
  }
  
  @IBAction func remove(_ sender: Any) {
    if (isAddPolygon){
      polygon?.map = nil
      isAddPolygon = false
      isInitPolygon = false
    }
  }
  
  @IBAction func hide(_ sender: Any) {
    polygon?.visible = !(polygon?.visible)!
    if ((polygon?.visible)!){
      lbHide.setTitle("Hide", for: .normal)
    } else {
      lbHide.setTitle("Show", for: .normal)
    }
  }
  @IBAction func update(_ sender: Any) {
    if (isAddPolygon){
        // update path
        // remove last coordinate
        pathPolygon.removeLastCoordinate()
        // add new coordinatre
        pathPolygon.add(CLLocationCoordinate2DMake(16.046232,108.194873))
        // add first coordinate
        pathPolygon.add(pathPolygon.object(at: 0))
        polygon?.path = pathPolygon
        // update color
        polygon?.fillColor = .red
    }
  }
}


extension PolygonViewController: MFMapViewDelegate {
  func mapview(_ mapView: MFMapView!, didTap polygon: MFPolygon!) {
    lbTitle.text = "Polygon clicked Id= " + String(polygon.id)
  }
  func mapView(_ mapView: MFMapView!, didTapAt coordinate: CLLocationCoordinate2D) {
    lbTitle.text = "Map clicked: " + String(coordinate.latitude) + " , " + String(coordinate.longitude)
  }
}
